server/src/main/java/org/elasticsearch/repositories/RepositoriesStatsCollector.java

```diff
@@ -0,0 +1,155 @@
+
+    private static final Logger logger = LogManager.getLogger(RepositoriesStatsCollector.class);
+
+    public static final TimeValue MIN_INTERVAL = TimeValue.timeValueSeconds(1L);
+
+    public static final Setting<TimeValue> INTERVAL =
+        Setting.timeSetting("repositories.monitoring.export.interval", TimeValue.timeValueMinutes(1L), MIN_INTERVAL,
+            Setting.Property.Dynamic, Setting.Property.NodeScope);
+
+    public static final Setting<Boolean> ENABLED =
+        Setting.boolSetting("repositories.monitoring.export.enabled", true,
+            Setting.Property.Dynamic, Setting.Property.NodeScope);
+
+    private final StatsCollector statsCollector = new StatsCollector();
+
+    private final Supplier<Map<String, RepositoryStats>> statsProvider;
+    private final ThreadPool threadPool;
+    private final Consumer<Map<String, RepositoryStats>> exporter;
+    private final ClusterService clusterService;
+    private volatile TimeValue interval;
+    private volatile boolean enabled;
+    private volatile Scheduler.Cancellable scheduler;
+
+    public RepositoriesStatsCollector(final Settings settings,
+                                      final ClusterService clusterService,
+                                      final Supplier<Map<String, RepositoryStats>> statsSupplier,
+                                      final ThreadPool threadPool) {
+        this(settings, clusterService, statsSupplier, threadPool, RepositoriesStatsCollector::exportMetrics);
+    }
+
+    public RepositoriesStatsCollector(final Settings settings,
+                                      final ClusterService clusterService,
+                                      final Supplier<Map<String, RepositoryStats>> statsSupplier,
+                                      final ThreadPool threadPool,
+                                      final Consumer<Map<String, RepositoryStats>> exporter) {
+        this.statsProvider = statsSupplier;
+        this.threadPool = threadPool;
+        this.exporter = exporter;
+        this.interval = INTERVAL.get(settings);
+        this.enabled = ENABLED.get(settings);
+        this.clusterService = clusterService;
+    }
+
+    private static void exportMetrics(Map<String, RepositoryStats> repositoriesStats) {
+        logger.info("Repositories stats {}", repositoriesStats);
+    }
+
+    private String threadPoolName() {
+        return ThreadPool.Names.GENERIC;
+    }
+
+    @Override
+    protected void doStart() {
+        scheduleStatsCollection();
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(ENABLED, this::setEnabled);
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(INTERVAL, this::setInterval);
+    }
+
+    @Override
+    protected void doStop() {
+        if (scheduler != null) {
+            cancelScheduledCollection();
+        }
+    }
+
+    @Override
+    protected void doClose() {}
+
+    private void scheduleStatsCollection() {
+        if (scheduler != null) {
```

Comment:
Do we need to synchronize here? It seems like we could have races between a setting update and stopping this component where we do non-atomic things to the `scheduler`?

Reply:
You're totally right. There's a chance of a race condition there.